### ISSUE TITLE
Speed up the expanded benchmark workflow

### ISSUE THREAD

**Issue by mattmccutchen-cci** [correctcomputation/actions#8]
With #7, the workflow takes over an hour to run.  Some ideas to speed it up:

- Use a release build of 3C and Clang, or RelWithDebInfo if having the debug info is useful (e.g., if it shows up in stack traces for assertion failures: I'd need to re-test that).
- For the CMake projects, replace `make` with `ninja`.  For the non-CMake projects, manually pass the appropriate flags to `make` for a parallel build.
- Generate the compilation database without actually building the unconverted code.
  - For the CMake projects, we should just be able to use `CMAKE_EXPORT_COMPILE_COMMANDS`, although I don't know if `convert_project` handles this correctly; it may need minor changes.
  - For the non-CMake projects, we could try to use [Bear's "wrapper" mode](https://github.com/rizsotto/Bear/wiki/Config-file#intercept) that uses a fake compiler that doesn't actually perform compilation, but this appears to require a newer version of Bear than is currently available via Ubuntu; we'd have to manually build our own version on gamera.

**Comment by john-h-kastner** [comment]
I originally switch from a release build to the debug build because assertions aren't included in the release, but I now see that `-DLLVM_ENABLE_ASSERTIONS=ON` exists. With this flag I think going back to a release build is fine.

**Comment by mattmccutchen-cci** [comment]
We've made a lot of progress here, but further speedups would help us further improve our development velocity with the current workflow and make it less painful to add more stuff to the workflow in the future.  Some more ideas for speedups:
- [ ] Use incremental rather than clean builds of the `checkedc-clang` repository in at least some cases.  Obviously, this introduces some risk of incorrect builds due to unsoundness in the build system; I know some people in the industry consider the use of clean builds in contexts like this sacrosanct.  But depending on how we implement it (see below), I think this risk is probably well worth accepting in order to reduce what is currently 12 minutes at the beginning of each run that we haven't been able to do anything about to perhaps 1-2 minutes in the common case.  Some possible approaches:
  1. Just don't have the workflow delete the source and build directories, so subsequent builds will be incremental by default.  If we suspect something is broken, we can log in and delete the directories manually, or we could have an option to do this on a manually triggered workflow run.
  2. Devise a means of quickly backing up and restoring the source and build directories that preserves all information needed for incremental builds to work (mainly, file modification times?); `rsync -a --del` might be adequate for this.  Then add a "clean build" flag to the workflow.  If the flag is on, the workflow does a clean build and backs up the result; if it is off, the workflow restores the backup and then does an incremental build.  For scheduled workflow runs, the flag would be on (because no one should be waiting to see workflow results in the middle of the night); for manually triggered runs, the flag would be off by default but could be turned on if we suspect the backed-up state is broken.  Compared to approach (i), this would help prevent buggy manually triggered runs from affecting subsequent runs at what I expect would be a modest additional cost for the backup and restore.
- [ ] For each benchmark, do the pre-conversion build once and reuse the result for all variants (`-alltypes` and not, and anticipated in the near future, macro-expanded and not).  This will probably shave a few minutes off the running time: ignoring the overhead of the mechanism itself, we'd save (number of variants per benchmark - 1) * (total pre-conversion build time of all benchmarks), and I think the latter factor is around a minute (libarchive alone is ~30 seconds).  Obviously, this reuse becomes obsolete if we find a way to skip the pre-conversion build altogether as discussed in my previous comment, but if we find reuse to be significantly easier to implement, it might be worth doing in the interim.

    We can't simply copy the directory because build systems tend to record absolute paths; in theory, this problem shouldn't exist if the entire industry used relative paths and did the hard work to track their proper interpretation in all contexts, but that's not the world we live in.  But assuming the variants never run in parallel (which seems to be the case now), a hack that might work is to back up the pre-conversion build directory and then have each variant restore it to the original absolute path, run the rest of the steps, and save the results to a variant-specific directory so we can inspect them later by logging in.  Alternatively, we may be able to do tricks with Linux mount namespaces to make different variants see different physical directories at the same absolute path.

Opinions (especially on the first idea; the second seems lower priority)?
